src/ObjWriting/XModel/Export/XModelBinWriter.cpp

```diff
@@ -0,0 +1,484 @@
+            XModelBinWriterBase::Write(material.incandescence[3]);
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_COEFFS);
+            XModelBinWriterBase::Write(material.coeffs[0]);
+            XModelBinWriterBase::Write(material.coeffs[1]);
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_GLOW);
+            XModelBinWriterBase::Write(material.glow.x);
+            XModelBinWriterBase::Write(material.glow.y);
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_REFRACTIVE);
+            XModelBinWriterBase::Write(material.refractive.x);
+            XModelBinWriterBase::Write(material.refractive.y);
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_SPECULAR_COLOR);
+            XModelBinWriterBase::Write(material.specularColor[0]); // R
+            XModelBinWriterBase::Write(material.specularColor[1]); // G
+            XModelBinWriterBase::Write(material.specularColor[2]); // B
+            XModelBinWriterBase::Write(material.specularColor[3]); // A
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_REFLECTIVE_COLOR);
+            XModelBinWriterBase::Write(material.reflectiveColor[0]); // R
+            XModelBinWriterBase::Write(material.reflectiveColor[1]); // G
+            XModelBinWriterBase::Write(material.reflectiveColor[2]); // B
+            XModelBinWriterBase::Write(material.reflectiveColor[3]); // A
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_REFLECTIVE);
+            XModelBinWriterBase::Write(material.reflective.x);
+            XModelBinWriterBase::Write(material.reflective.y);
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_BLINN);
+            XModelBinWriterBase::Write(material.blinn[0]);
+            XModelBinWriterBase::Write(material.blinn[1]);
+
+            XModelBinWriterBase::Write(XModelBinHash::MATERIAL_PHONG);
+            XModelBinWriterBase::Write(material.phong);
+
+            materialNum++;
+        }
+    }
+
+public:
+    XModelBinWriter7(std::ostream& stream, std::string gameName, std::string zoneName)
+        : XModelBinWriterBase(stream, std::move(gameName), std::move(zoneName))
+    {
+    }
+
+    void Write(const XModelCommon& xmodel) override
+    {
+        PrepareVertexMerger(xmodel);
+        WriteHeader(7);
+        WriteBones(xmodel);
+        WriteVertices(xmodel);
+        WriteFaces(xmodel);
+        WriteObjects(xmodel);
+        WriteMaterials(xmodel);
+
+        auto estimatedCompressedFileSize = LZ4_compressBound(m_writer.str().size());
+        auto compressedBuffer = new char[estimatedCompressedFileSize];
+        auto actualCompressedFileSize = LZ4_compress_default(m_writer.str().c_str(), compressedBuffer, m_writer.str().size(), estimatedCompressedFileSize);
+
+        auto uncompressedSize = m_writer.str().size();
+        char uncompressedSizeChar[4];
+        std::memcpy(uncompressedSizeChar, &uncompressedSize, sizeof(uncompressedSizeChar));
```

Comment:
Doing this is a bit dangerous here, the returned `size()` is a `size_t` which is 4 bytes in 32bit but 8bytes in 64bit.
Since the variable is `auto` type, it will get bigger here as well.

Additionally, you dont need another variable here since you can cast the pointer to the numeric variable.

```cpp
auto uncompressedSize = static_cast<uint32_t>(m_writer.str().size());
// ....
m_stream.write(reinterpret_cast<char*>(&uncompressedSize), sizeof(uncompressedSize));
```